tl.tl

```diff
@@ -6383,17 +6414,29 @@ tl.type_check = function(ast: Node, opts: TypeCheckOptions): {Error}, {Error}, T
                   node_error(node, "cannot determine type of table literal")
                end
             elseif is_array then
-               for i = 2, #node.type do
-                  if not is_a(node.type[i], node.type[i-1]) then
-                     pure_array = false
-                     break
+               local pure_array = true
+               local last_idx = 1
+               for i = 2, largest_array_idx do
+                  if not node.type[i] then
+                     -- continue
```

Comment:
I'm failing to understand this part. What is understood as a "pure array" here?

Some potentially clarifying questions:

* Can/should the table be inferred as tupletable if the types are not sequentially present? e.g. `{ [3] = true }`.

* Another concern that came to mind when I saw this loop which I think may be a sign of something possibly problematic: what if the key has a huge index like this: `{ [2^50] = true }`. Might want to pairs over the present items processing the positive indices (which I think are the only ones stored?) rather than ipairs over all possible indices.

* Is the `is_a` test there to check that all elements have the same type? If so, note that comparisons with `is_a` here may end up causing weird side-effects between the order of elements in the table literal and subtypes, since you're running `is_a` against the last index each time (this may match positive for a tower of subtypes, and then fail if you invert the order of the elements of the array). `same_type` would be a safer bet.

Reply:
A 'pure array' in this context is an array of a type that isn't a union basically (pure was the only concise word I could really think of). 

> Can/should the table be inferred as tupletable if the types are not sequentially present? e.g. { [3] = true }

I think that is currently an array? At least running 
```
print_type{ [3] = true }
```
prints out `{boolean}`
At least in my use case, I've been writing sparse tuples with an explicit `nil`, like `{ nil, nil, true }`, which is inferred as a tuple.
But I think that should probably stay as an array?

> Might want to pairs over the present items processing the positive indices [...] rather than ipairs over all possible indices

:+1:

> Is the is_a test there to check that all elements have the same type?

Yeah

> `same_type` would be a safer bet

:+1:

Reply:
Alright, most recent commit uses `pairs` over the existing indices and `same_type` instead of `is_a`